### PR TITLE
#308: Improve BC file formatting

### LINKED ISSUE #308: Improve bc file formatting
**Is your feature request related to a problem? Please describe.**
example code:
```
from pathlib import Path

from hydrolib.core.io.bc.models import (
    ForcingModel,
    QuantityUnitPair,
    TimeInterpolation,
    TimeSeries,
)

# Chosen approach: separate .bc file for each boundary (east and south)
bc_east = ForcingModel()

# Each .bc file can contain 1 or more timeseries, one for each support point:
steric = TimeSeries(
    name="east2_0001",
    quantityunitpair=[QuantityUnitPair(quantity="time", unit="seconds since 2022-01-01 00:00:00 +00:00"), #TODO: should fail if only one QUP is supplied, but does not: https://github.com/Deltares/HYDROLIB-core/issues/305
                      QuantityUnitPair(quantity="waterlevel", unit="m")],
    timeinterpolation=TimeInterpolation.linear,
    datablock=[[0, 1.1], [3600, 2.3], [7200, 3.5], [10800, 2.4], [86400, -0.123]],
)
bc_east.forcing.append(steric)

bc_east.save(filepath=Path("steric_east2.bc"))
```

Results in this file:
```
[Forcing]
name              = east2_0001
function          = timeseries
timeInterpolation = linear
offset            = 0.0
factor            = 1.0
quantity          = time
unit              = seconds since 2022-01-01 00:00:00 +00:00
quantity          = waterlevel
unit              = m
0.0        1.1
3600.0     2.3
7200.0     3.5
10800.0    2.4
86400.0    -0.123 
```

**Describe the solution you'd like**
The time and data values are not neatly aligned. This can easily be solved. User provides a formatting string for time and data columns, or done semi-automatically. Some thinking is still required, since different data types require different precision.

**Describe alternatives you've considered**
A clear and concise description of any alternative solutions or features you've considered (if applicable).

**Additional context**
This is important with large t3D bc files, because of readability, but also disk space. The latter is significantly reduced when reducing precision and that is relevant since even then the bc files are >1GB. It could also be that formatting with limited significance improves writing speed.

### PR DESCRIPTION
**Pull request by priscavdsluis**
Fixes #308

### CODE REVIEW

**Comment by priscavdsluis** [comment]
@veenstrajelmer, 
I made an initial setup. Would you please check it out and see if this works for you?
Note that the functionality in this branch, is only functional for INI files (so no polyfiles yet).
I will work on that next, if I have a green light from you.

### **1. For all INI files it is possible to do the following:**
```python
ext_model = ExtModel(filepath="some_file_path.ext")
ext_model.serializer_config.number_of_decimals = 3
ext_model.save()
```

All float values of the properties will be written with 3 decimals.

### **2. For all INI files that have a datablock (ForcingModel) it is possible to do the following:**
```python
forcing_model = ExtModel(filepath="some_file_path.bc")
forcing_model.serializer_config.number_of_decimals = 3
forcing_model.serializer_config.number_of_decimals_datablock = 2
forcing_model.save()
```

All float values of the properties will be written with 3 decimals.
All float values inside the datablock will be written with 2 decimals.

### **3. Additional notes**
- These settings are by default not set (None).
- If these settings or not set, the full original float value will be used.
- The setting does not influence the data in the data model; it is only used for serialization.
- The current implementation does not have breaking changes.

**Comment by priscavdsluis** [comment]
Note that I have created a separate PR for the rest of the files: #417

**Comment by priscavdsluis** [comment]
I have decided to keep formatting the time columns: I had an initial implementation for it which was small enough, but then I noticed that for example a harmonic forcing has "minutes" as unit of the harmonic component. We might want to create a separate issue for this and think about the design. I believe it is out of scope for now.

File: hydrolib/core/io/rr/models.py
Comment (tim-vd-aardweg) [inline hydrolib/core/io/rr/models.py:517]: Perhaps update the comment why we are now skipping the first 2?

**Review by tim-vd-aardweg — COMMENTED** [review]
As requested, an initial review focused on the bigger picture instead of the details. Let's discuss this tomorrow.

File: hydrolib/core/basemodel.py
Comment (tim-vd-aardweg) [reply hydrolib/core/basemodel.py:1091]: For now it's no problem! :)

File: hydrolib/core/io/dflowfm/bc/models.py
Comment (tim-vd-aardweg) [reply hydrolib/core/io/dflowfm/bc/models.py:784]: Yes!

File: hydrolib/core/io/dflowfm/ini/models.py
Comment (tim-vd-aardweg) [reply hydrolib/core/io/dflowfm/ini/models.py:260]: This one actually needs to call the `write_ini()` so it cannot be removed.

**Comment by priscavdsluis** [comment]
@veenstrajelmer, I also did some performance testing with the script that you gave me.
I run the tests for each version 3 times. 
I also made the values in the data block uniform by replacing
```python
datablock = np.random.uniform(low=-40, high=130.3, size=(nrows,50))
```
with
```python
datablock = np.full((nrows, 50), 1.23)datablock = np.full((nrows, 50), 1.23)
```
These are my results:

```
---FEAT/308-IMPROVE-BC-FILE-FORMATTING (1f3dbfe, 17-11-22)
1000 rows datablock generation:               	0.14 sec  	0.13 sec	0.13 sec	
1000 rows with hydrolib (unformatted):        	0.28 sec  	0.25 sec    0.27 sec
1000 rows with hydrolib (formatted single):   	0.30 sec  	0.26 sec    0.30 sec
1000 rows with savetxt (unformatted):         	0.08 sec  	0.11 sec    0.08 sec
1000 rows with savetxt (formatted single):    	0.03 sec  	0.04 sec    0.04 sec
1000 rows with savetxt (formatted percol):    	0.03 sec  	0.03 sec    0.03 sec
																			
10000 rows datablock generation:              	1.38 sec  	1.40 sec    1.53 sec
10000 rows with hydrolib (unformatted):       	2.89 sec  	2.93 sec    3.12 sec
10000 rows with hydrolib (formatted single):  	3.14 sec  	3.10 sec    3.09 sec
10000 rows with savetxt (unformatted):        	0.68 sec  	0.65 sec    0.70 sec
10000 rows with savetxt (formatted single):   	0.23 sec  	0.23 sec    0.26 sec
10000 rows with savetxt (formatted percol):   	0.24 sec  	0.22 sec    0.21 sec
                                                                    
100000 rows datablock generation:             	       13.14 sec 	13.19 sec   13.63 sec
100000 rows with hydrolib (unformatted):      	30.16 sec 	30.11 sec   32.23 sec
100000 rows with hydrolib (formatted single):	29.66 sec 	30.90 sec   31.85 sec
100000 rows with savetxt (unformatted):       	7.02 sec  	6.73 sec    7.29 sec
100000 rows with savetxt (formatted single):  	2.40 sec  	2.32 sec    2.40 sec
100000 rows with savetxt (formatted percol):  	2.28 sec  	2.08 sec    2.19 sec


---MAIN (aed6dc1, 17-11-22)
1000 rows datablock generation:             	0.15 sec	0.13 sec	0.16 sec	
1000 rows with hydrolib (unformatted):      	0.27 sec    0.24 sec    0.24 sec
1000 rows with hydrolib (skipped):
1000 rows with savetxt (unformatted):       	0.08 sec    0.08 sec    0.08 sec
1000 rows with savetxt (formatted single):  	0.03 sec    0.03 sec    0.03 sec
1000 rows with savetxt (formatted percol):  	0.02 sec    0.03 sec    0.03 sec
                                                                        
10000 rows datablock generation:             	1.34 sec    1.39 sec    1.58 sec
10000 rows with hydrolib (unformatted):      	2.78 sec    2.91 sec    2.75 sec
10000 rows with hydrolib (skipped):
10000 rows with savetxt (unformatted):       	0.70 sec    0.74 sec    0.69 sec
10000 rows with savetxt (formatted single):  	0.27 sec    0.25 sec    0.24 sec
10000 rows with savetxt (formatted percol):  	0.25 sec    0.22 sec    0.23 sec
                                                                        
100000 rows datablock generation:             	12.96 sec   13.89 sec   12.83 sec
100000 rows with hydrolib (unformatted):      	29.64 sec   31.81 sec   29.57 sec
100000 rows with hydrolib (skipped):
100000 rows with savetxt (unformatted):       	7.00 sec    7.25 sec    6.97 sec
100000 rows with savetxt (formatted single):  	2.29 sec    2.29 sec    2.38 sec
100000 rows with savetxt (formatted percol):  	2.11 sec    2.22 sec    2.14 sec


---0.3.0
1000 rows datablock generation:             	0.12 sec	0.12 sec 	0.15 sec
1000 rows with hydrolib (unformatted):      	0.25 sec    0.23 sec 	0.25 sec
1000 rows with hydrolib (skipped): 				-           -        	-
1000 rows with savetxt (unformatted):       	0.09 sec    0.10 sec 	0.10 sec
1000 rows with savetxt (formatted single):  	0.03 sec    0.03 sec 	0.03 sec
1000 rows with savetxt (formatted percol):  	0.04 sec    0.02 sec 	0.03 sec
																		
10000 rows datablock generation:             	1.43 sec    1.32 sec 	1.44 sec
10000 rows with hydrolib (unformatted):      	3.03 sec    2.77 sec 	2.85 sec
10000 rows with hydrolib (skipped): 			-           -        	-
10000 rows with savetxt (unformatted):       	0.76 sec    0.73 sec 	0.66 sec
10000 rows with savetxt (formatted single):  	0.29 sec    0.22 sec 	0.23 sec
10000 rows with savetxt (formatted percol):  	0.23 sec    0.20 sec 	0.24 sec
																		
100000 rows datablock generation:             	13.41 sec   13.27 sec	13.06 sec
100000 rows with hydrolib (unformatted):      	31.75 sec   30.56 sec	30.25 sec
100000 rows with hydrolib (skipped): 			-           -        	-
100000 rows with savetxt (unformatted):       	7.08 sec    7.11 sec 	6.85 sec
100000 rows with savetxt (formatted single):  	2.40 sec    2.31 sec 	2.59 sec
100000 rows with savetxt (formatted percol):  	2.28 sec    2.32 sec 	2.19 sec
```

I currently see no significant changes between the main, the branch and 0.3.1, so there is no performance regression in this branch.
To improve the performance further, I will create a separate issue.

File: hydrolib/core/io/dflowfm/ini/serializer.py
Comment (priscavdsluis) [reply hydrolib/core/io/dflowfm/ini/serializer.py:60]: I agree with this comment. 
However, this would require a bigger refactoring.
The SectionSerializer and the Serializer both use the INISerializerConfig and use its data block settings. To refactor this we should have a Serializer for section with data blocks. 
So unfortunately, I will skip this for now and create a new issue.

File: hydrolib/core/io/rr/models.py
Comment (priscavdsluis) [reply hydrolib/core/io/rr/models.py]: I have updated the comment.

File: hydrolib/core/io/dflowfm/ini/serializer.py
Comment (priscavdsluis) [reply hydrolib/core/io/dflowfm/ini/serializer.py:60]: Created issue: #424

**Review by tim-vd-aardweg — COMMENTED** [review]
Just a couple small remarks I'd like to discuss before merging back! :)

File: hydrolib/core/io/dflowfm/bc/models.py
Comment (priscavdsluis) [reply hydrolib/core/io/dflowfm/bc/models.py:759]: Yes have added it.

File: hydrolib/core/io/dflowfm/ext/models.py
Comment (priscavdsluis) [reply hydrolib/core/io/dflowfm/ext/models.py:206]: Added it.

**Review by tim-vd-aardweg — COMMENTED** [review]
These changes are good!

**Review by tim-vd-aardweg — APPROVED** [review]

**Pull request merged**
